### ISSUE TITLE
RejectLogger setup and File Read Error policy

### ISSUE THREAD

**Issue by ninjapapa** [TresAmigosSD/SMV#160]
So far all of our use cases for RejectLogger are to handle file reading errors, which could be either in the step of 
* parsing the entire string of record to sequence of string, or
* covert strings to values

We need 3 policies on file reading handling on file level, i.e. different files could have different policies. The 3 policies are:
* Ignore - just drop error records and continue reading next record
* Terminate - for whatever error, terminate the reading process and print the record which caused the error
* Log - put the error record in a logger and keep going. This one we need to output the content of the logger when reading is done

We already have the 3 policies as parameter to SmvFile, however, there are 2 things current Reject logger does not support yet:
* Terminate with print
* Check the logger and dump it when the reading is done

The tricky part is that reading happens at the action step, however SmvFile itself can't know what is the action. The closest place which guarantees an action is the persist step of each SmvModule, so we need to dump reject logger at each persist of SmvModule.

**Comment by AliTajeldin** [comment]
What would happen in the above scenario if someone initiated an action before persisting the file?  For example, if someone run a "collect()" on an rdd and then use the result to perform some operations on another DF.  Will we loose the collected logs up to that point or will they still be around the next time we do a module persist action?
If they always show up (perhaps with a slight delay), then it is no big deal.  If they do get dropped, then we need to rethink (perhaps output them on each executer rather than collect them and send them to the driver).

**Comment by ninjapapa** [comment]
They will not be dropped if we have a single app level logger. The disadvantage of a single app level logger is that we can't know file names from the rejected record. For example a module combine multiple input files, when we have a single app level logger and dump at the persisting step of the module, we can't tell which file has the data issue.

**Comment by AliTajeldin** [comment]
We should know at the time we encounter the error (in the parser) where the error occurred.  We must be in the context of an SmvCsvFile at that point.  The fact that we don't is an indication of bad design.

At the time I wrote the original CSV file handler, all the intermediate helper functions were attached to the actionable object.  So parsing the RDD was added to RDD helper and applying the schema was added to Schema Helper and so on.  That is great if one want to use these helper functions outside of the context of parsing a CSV file but that has proven otherwise.

We really should move these into the SmvCsvFile class and we would then know (or be passed in) the name of the file that we are operating on at the moment.  We can then utilize a global/app level reject logger as each error message would have the name of the offending file.

On a somewhat related note, I think having CsvAttribute and RejectLogger as implicit parameters was also a mistake.
* CsvAttribute is always needed and should be passed in as a regular parameter.  We can create helper factory methods that create files with common attributes instead.
* RejectLogger should be an optional parameter (type Option[RejectLogger[) that defaults to None.
